docs/GDPs/gdp-duck-storage.rst

```diff
@@ -0,0 +1,456 @@
+:code:`alignment`
+   the resulting alignment is chosen as the least common multiple of the alignments of all inputs which are a GT4Py
+   Storage
+:code:`dtype`
+   the resulting dtype is determined by NumPy behavior
+
+
+Mixing Types
+============
+
+If a binary ufunc is applied to a storage and a non-storage array, the storage determines the behavior.
+Since non-storage arrays do not carry the necessary information to apply the usual broadcasting rules,
+we only implement the cases where
+
+* the array has the same shape as the input storage or as the broadcast shape when considering a provided output buffer
+* the array has a 3d shape where dimensions with shape :code:`1` in the array are broadcast.
+
+Mixing Devices
+==============
+
+For the synchronized memory classes (be it by CUDA or by GT4Py), the compute device is chosen depending on
+
+:code:`CudaManagedGPUStorage`
+   The compute device is chosen to be GPU iff inputs are comptaible with `cp.ndarray`.
+
+:code:`SoftwareManagedGPUStorage`
+   Here, array is considered a GPU array if it is compatible with :code:`cp.asarray`. If a storage is modified on CPU,
+   it is considered a CPU array here. The compute device is chosen as GPU unless all inputs are not GPU arrays.
+   (including if all inputs are :code:`SoftwareManagedGPUStorage` but are modified on CPU)
+
+We assume that mixing these in the same application is not a common case. Should it nevertheless appear, the object that
+handles the ufunc will determine the behavior. (Where each of the classes will treat the other as on GPU.)
+
+For CPU storages, all inputs and output need to be compatible with `np.asarray`, for GPU storages with `cp.asarray`,
+otherwise an exception is raised.
+
+:code:`CudaManagedGPUStorage` and :code:`SoftwareManagedGPUStorage` shall both have a :code:`__array_priority__` set to
+:code:`11`, while for :code:`CPUStorage` and :code:`GPUStorage` it is set to :code:`10`, meaning that managed storages
+have priority in handling these cases.
+
+Annotation of Stencils
+^^^^^^^^^^^^^^^^^^^^^^
+
+Currently, field arguments are annotated with :code:`Field[dtype]` in the function signature. The assumed layout and
+alignment in the generated code is then based on the :code:`backend` parameter of the :code:`stencil` decorator.
+This will continue to work, but in case the storage passed at call-time uses other settings than the backend's default
+settings, these must also be specified to the stencil. We propose the following arguments for the :code:`Field`
+annotation, which are specified using the notation (:code:`Argument[value]`):
+
+:code:`DType`
+   correspoinds to the `dtype` argument
+:code:`LayoutMap`
+   corresponds to the `layout_map` argument
+:code:`Alignment`
+   corresponds to the `alignment` argument
+:code:`DefaultParameters`
+   corresponds to the `default_parameters` argument.
+   Either :code:`'F'` for FORTRAN layout, :code:`'C'` for C/C++-layout or one of the backend identifier strings.
+:code:`Axes`
```

Comment:
Why do we need to pass these informations to the stencil? What benefit would you get? We should separate concerns between data organization and algorithms... Or is it just to do runtime checks for issuing warnings and maybe errors?

Reply:
Are you asking only about Axes, or all of them? Axes needs to be there so you know how many dimensions the field has, which is needed for both analysis and to generate the GT code. Applies also to LayoutMap and Alignment.

Reply:
This part has been removed from the GDP.